examples/hmm.py

```diff
@@ -569,50 +574,56 @@ def main(args):
+        guide_trace = handlers.trace(guide).get_trace(
             sequences, lengths, args=args, batch_size=args.batch_size)
-        model_trace = poutine.trace(
-            poutine.replay(poutine.enum(model, first_available_dim), guide_trace)).get_trace(
+        model_trace = handlers.trace(
+            handlers.replay(handlers.enum(model, first_available_dim), guide_trace)).get_trace(
             sequences, lengths, args=args, batch_size=args.batch_size)
         logging.info(model_trace.format_shapes())
 
+    # Appease the PyTorch JIT
+    model_name = model.__name__
+    model = functools.partial(model, args=args)
+    guide = functools.partial(guide, args=args)
+
     # Enumeration requires a TraceEnum elbo and declaring the max_plate_nesting.
     # All of our models have two plates: "data" and "tones".
-    optim = Adam({'lr': args.learning_rate})
+    optimizer = optim.Adam({'lr': args.learning_rate})
     if args.tmc:
-        if args.jit:
+        if args.jit and not args.funsor:
             raise NotImplementedError("jit support not yet added for TraceTMC_ELBO")
-        elbo = TraceTMC_ELBO(max_plate_nesting=1 if model is model_0 else 2)
-        tmc_model = poutine.infer_config(
+        Elbo = infer.JitTraceTMC_ELBO if args.jit else infer.TraceTMC_ELBO
+        elbo = Elbo(max_plate_nesting=1 if model is model_0 else 2)
+        tmc_model = handlers.infer_config(
             model,
             lambda msg: {"num_samples": args.tmc_num_samples, "expand": False} if msg["infer"].get("enumerate", None) == "parallel" else {})  # noqa: E501
-        svi = SVI(tmc_model, guide, optim, elbo)
+        svi = infer.SVI(tmc_model, guide, optimizer, elbo)
     else:
-        Elbo = JitTraceEnum_ELBO if args.jit else TraceEnum_ELBO
+        Elbo = infer.JitTraceEnum_ELBO if args.jit else infer.TraceEnum_ELBO
         elbo = Elbo(max_plate_nesting=1 if model is model_0 else 2,
                     strict_enumeration_warning=(model is not model_7),
                     jit_options={"time_compilation": args.time_compilation})
-        svi = SVI(model, guide, optim, elbo)
+        svi = infer.SVI(model, guide, optimizer, elbo)
 
     # We'll train on small minibatches.
     logging.info('Step\tLoss')
     for step in range(args.num_steps):
-        loss = svi.step(sequences, lengths, args=args, batch_size=args.batch_size)
+        loss = svi.step(sequences, lengths, batch_size=args.batch_size)
         logging.info('{: >5d}\t{}'.format(step, loss / num_observations))
 
     if args.jit and args.time_compilation:
         logging.debug('time to compile: {} s.'.format(elbo._differentiable_loss.compile_time))
 
     # We evaluate on the entire training dataset,
     # excluding the prior term so our results are comparable across models.
-    train_loss = elbo.loss(model, guide, sequences, lengths, args, include_prior=False)
+    train_loss = elbo.loss(model, guide, sequences, lengths, batch_size=sequences.shape[0], include_prior=False)
```

Comment:
Why do default args no longer work?

Reply:
I'm not sure where exactly the failure starts but I think there are a bunch of places in Funsor and possibly `contrib.funsor` that convert shapes to tuples and integers and screw up PyTorch JIT shape inference, and this causes a shape error here and further down at `test_loss =` after JIT compilation.